lib/poseidon/sync_producer.rb

```diff
@@ -71,7 +72,8 @@ def send_messages(messages)
       end
 
       if messages_to_send.pending_messages?
-        raise "Poseidon: Failed to send all messages: #{messages_to_send.messages} remaining"
+        failed_messages = messages_to_send.messages
+        raise Errors::MessageFailedToSendError.new(failed_messages, "Poseidon: Failed to send #{failed_messages.size} out of #{total_message_count} messages")
```

Comment:
That looks pretty good to me.  Poseidon's error handling is beyond convoluted so I'll never be convinced we've caught everything :)  I'm good with this, though.